index.js

```diff
@@ -55,21 +55,54 @@ function findAllDependencies(file, knownDependencies, baseDir, knownFiles) {
     return knownDependencies;
   }
 
-  if (baseDir) {
-    return findAllDependenciesHelp(file, knownDependencies, baseDir, knownFiles).then(function(thing){
+  if (sourceDirectories) {
+    return findAllDependenciesHelp(file, knownDependencies, sourceDirectories, knownFiles).then(function(thing){
       return thing.knownDependencies;
     });
   } else {
-    return getBaseDir(file).then(function(newBaseDir) {
-      return findAllDependenciesHelp(file, knownDependencies, newBaseDir, knownFiles).then(function(thing){
-        return thing.knownDependencies;
+    return getBaseDir(file)
+      .then(findElmPackage)
+      .then(getSourceDirectories)
+      .then(function(newSourceDirs) {
+        return findAllDependenciesHelp(file, knownDependencies, newSourceDirs, knownFiles).then(function(thing){
+          return thing.knownDependencies;
+        });
       });
-    })
   }
 }
 
+// Given a source directory (containing top-level Elm modules), locate the
+// elm-package.json file that includes it.
+function findElmPackage(baseDir, currentDir) {
```

Comment:
My intent here was to preserve `baseDir` through the recursion so that when we found the elm-package.json file, we could verify that the `baseDir` in question was actually listed in its `source-directories`, but I ended up not writing this check. What do you think? Would this level of robustness be useful? Or is it sufficient just to search up through parent directories and use the first elm-package.json file that we find (in which case I will delete the unneeded variable).